digest/src/newtype.rs

```diff
@@ -0,0 +1,528 @@
+        }
+    }
+
+    #[macro_export]
+    macro_rules! newtype_common {
+        ($name: ident, $inner: ty) => {
+            $crate::newtype_common_customized_init!($name, $inner);
+            $crate::newtype_common_extendable_output!($name, $inner);
+            $crate::newtype_common_extendable_output_reset!($name, $inner);
+            $crate::newtype_common_fixed_output!($name, $inner);
+            $crate::newtype_common_fixed_output_reset!($name, $inner);
+            $crate::newtype_common_hash_marker!($name, $inner);
+            $crate::newtype_common_inner_user!($name, $inner);
+            $crate::newtype_common_output_size_user!($name, $inner);
+            $crate::newtype_common_reset!($name, $inner);
+            $crate::newtype_common_update!($name, $inner);
+            $crate::newtype_common_serializable_state!($name, $inner);
+        };
+    }
+}
+
+#[cfg(feature = "mac")]
+mod mac {
+    use super::Wrapper;
+    use crate::{InnerInit, MacMarker};
+    use core::marker::PhantomData;
+
+    impl<T> InnerInit for Wrapper<'_, T>
+    where
+        T: InnerInit,
+    {
+        fn inner_init(inner: Self::Inner) -> Self {
+            Wrapper {
+                inner: T::inner_init(inner),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    impl<T> MacMarker for Wrapper<'_, T> where T: MacMarker {}
+
+    #[macro_export]
+    macro_rules! newtype_mac {
+        ($name: ident, $inner: ty) => {
+            impl $crate::InnerInit for $name
+            where
+                for<'a> $crate::newtype::Wrapper<'a, $inner>: $crate::InnerInit,
+            {
+                #[inline]
+                fn inner_init(inner: Self::Inner) -> Self {
+                    Self($crate::newtype::Wrapper::<'static, $inner>::inner_init(
+                        inner,
+                    ))
+                }
+            }
+        };
+    }
+}
+
+#[cfg(not(feature = "mac"))]
+#[macro_export]
+macro_rules! newtype_mac {
+    ($name: ident, $inner: ty) => {};
+}
+
+#[cfg(feature = "oid")]
+mod oid {
+    use super::Wrapper;
+    use crate::const_oid::{AssociatedOid, ObjectIdentifier};
+
+    impl<T> AssociatedOid for Wrapper<'_, T>
+    where
+        T: AssociatedOid,
+    {
+        const OID: ObjectIdentifier = T::OID;
+    }
+
+    #[macro_export]
+    macro_rules! newtype_oid {
+        ($name: ident, $inner: ty) => {
+            impl $crate::const_oid::AssociatedOid for $name
+            where
+                for<'a> $crate::newtype::Wrapper<'a, $inner>: $crate::const_oid::AssociatedOid,
+            {
+                const OID: $crate::const_oid::ObjectIdentifier =
+                    $crate::newtype::Wrapper::<'static, $inner>::OID;
+            }
+        };
+    }
+}
+
+#[cfg(not(feature = "oid"))]
+#[macro_export]
+macro_rules! newtype_oid {
+    ($name: ident, $inner: ty) => {};
+}
+
+#[macro_export]
+macro_rules! newtype {
+    ($doc: expr, $name: ident = $inner: ty) => {
+        #[doc = $doc]
```

Comment:
It's better to pass docs like this:
```rust
macro_rules! newtype {
    ($(#[$attr:meta])* $name:ident) => {
        $(#[$attr])*
        pub struct $name {}
    };
}
```

It not only will allow to pass docs in a nicer way (instead of the `r"..."` abomination), but also will allow to `derive` traits if necessary:
```rust
newtype!(
    /// Docs
    #[derive(Clone)]
    Foo
);
```

Reply:
Agreed :) I could not find the correct type for the parameter of the macro!